### ISSUE TITLE
Add value validation

### ISSUE THREAD

**Issue by ajtritt** [hdmf-dev/hdmf#358]
Provide a generic way of validating contents of Containers/Datas.

**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Ex. I'm trying to achieve [...]

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

## Checklist

- [ ] Have you ensured the feature or change was not already [reported](https://github.com/hdmf-dev/hdmf/issues) ?
- [ ] Have you included a brief and descriptive title?
- [ ] Have you included a clear description of the problem you are trying to solve?
- [ ] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [ ] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?

**Comment by oruebel** [comment]
For NWB specifically, https://github.com/NeurodataWithoutBorders/nwbinspector may already define some of the cases relevant to value validation. While this is not relevant on this issue for HDMF it is likely relevant when using the HDMF value validation tools in PyNWB downstream.

**Comment by rly** [comment]
When this is addressed, it should include a check to make sure `DynamicTableRegion` data points to valid row indices of the target table. This was removed in https://github.com/hdmf-dev/hdmf/pull/390 for performance reasons, but is a worthwhile check somewhere, because it is possible to create and write a `DynamicTableRegion` with invalid row indices, which on read, cannot be accessed properly. 

For example:

```python
# [snip] create nwbfile and add 2 electrodes

nwbfile.add_unit(spike_times=[1, 2, 3], electrodes=[10000])

with NWBHDF5IO(filename, 'w') as io:
    io.write(nwbfile)

with NWBHDF5IO(filename, 'r') as io:
    nwbfile = io.read()
    print(nwbfile.units[0])  # raises ValueError: Index (10000) out of range (0-1)
```

**Comment by mavaylon1** [comment]
Closing due to this being the same issue/fix as #509
